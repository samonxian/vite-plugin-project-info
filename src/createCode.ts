import path from 'path';
import moment from 'moment';
import fs from 'fs';
import { ProjectInfoPluginOptions } from '.';

export const DEFAULT_LOCALE = {
  projectVersion: '项目版本',
  buildTime: '构建时间',
  projectName: '项目名称',
  projectDescription: '项目描述',
  projectAuthor: '负 责 人',
  repositoryLink: '仓库链接',
};

export const EN_LOCALE = {
  projectVersion: 'ProjectVersion',
  buildTime: 'BuildTime',
  projectName: 'ProjectName',
  projectDescription: 'ProjectDescription',
  projectAuthor: 'ProjectAuthor',
  repositoryLink: 'RepositoryLink',
};

export function createCode(opts: { locale?: ProjectInfoPluginOptions['locale'] } = {}) {
  const { locale = DEFAULT_LOCALE } = opts;
  const packageJson = require(path.resolve('package.json'));
  let { repository } = packageJson;

  if (typeof repository === 'object') {
    packageJson.repository = repository.url;
  }

  packageJson.buildTime = moment().format('YYYY-MM-DD HH:mm:ss');

  let versionCode = fs.readFileSync(path.resolve(__dirname, './clientOutputTemplate.tpl'), { encoding: 'utf8' });
  const pkgMatchs = versionCode.match(/{pkg\.[^{}]*}/g);
  pkgMatchs?.forEach((m) => {
    const fieldName = m.match(/{pkg\.(.*)}/)?.[1];
    versionCode = versionCode.replace(new RegExp(m, 'g'), packageJson[fieldName || ''] || '');
  });

  const localeMatchs = versionCode.match(/{locale\.[^{}]*}/g);
  localeMatchs?.forEach((m) => {
    const localeFieldName = m.match(/{locale\.(.*)}/)?.[1];
    versionCode = versionCode.replace(new RegExp(m, 'g'), locale[localeFieldName] || '');
  });

  return versionCode;
}
