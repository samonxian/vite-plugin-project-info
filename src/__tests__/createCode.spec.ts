import { describe, it, expect } from 'vitest';
import createCode from '../createCode';

describe('createCode', () => {
  it('create code correctly', () => {
    const code = createCode();
    expect(code).toContain("repository: 'https://github.com/samonxian/vite-plugin-project-info'");
    expect(code).toContain("version: '0.1.0'");
    expect(code).toContain("name: 'vite-plugin-project-info'");
    expect(code).toMatchSnapshot(code);
  });
});
