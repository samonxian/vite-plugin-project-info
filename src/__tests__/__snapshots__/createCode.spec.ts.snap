// Vitest Snapshot v1

exports[`createCode > create code correctly > const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:41:47',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:41:47', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
 1`] = `
"const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:41:47',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:41:47', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
"
`;

exports[`createCode > create code correctly > const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:42:04',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:42:04', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
 1`] = `
"const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:42:04',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:42:04', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
"
`;

exports[`createCode > create code correctly > const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:42:46',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:42:46', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
 1`] = `
"const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:42:46',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:42:46', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
"
`;

exports[`createCode > create code correctly > const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:44:23',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:44:23', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
 1`] = `
"const COLORS = {
  primary: '#1890ff',
  success: '#52c41a',
  info: '#13c2c2',
  danger: '#f5222d',
  // warning: '#fa8c16',
  // default: '#35495E',
};
/**
 * 固定模板的 console.log 输出
 * @param title log的标题
 * @param description log的描述
 * @param color 颜色
 */
function log(title, description, color) {
  if (title && description) {
    console.log(
      \`%c \${title} %c \${description} %c\`,
      \`background:\${color};border:1px solid \${color}; padding: 5px; border-radius: 4px 0 0 4px; color: #fff;\`,
      \`border:1px solid \${color}; padding: 5px; border-radius: 0 4px 4px 0; color: \${color};\`,
      'background:transparent',
    );
  }
}

const projectInfo = {
  name: 'vite-plugin-project-info',
  version: '0.1.0',
  name: '',
  repository: 'https://github.com/samonxian/vite-plugin-project-info',
  author: 'samonxian',
  buildTime: '2022-07-19 11:44:23',
  consoleLogProjectInfo() {
    log('项目版本', '0.1.0', '#eb2f96');
    log('构建时间', '2022-07-19 11:44:23', COLORS.danger);
    log('项目名称', 'vite-plugin-project-info', COLORS.primary);
    log('项目描述', '', '#722ed1');
    log('项目链接', 'https://github.com/samonxian/vite-plugin-project-info', COLORS.success);
    log('负 责 人', 'samonxian', COLORS.info);
  },
};

projectInfo.consoleLogProjectInfo();

export default projectInfo;
"
`;
